core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -798,7 +798,9 @@ private[spark] class ExecutorAllocationManager(
         }
         if (taskEnd.taskInfo.speculative) {
           stageAttemptToSpeculativeTaskIndices.get(stageAttempt).foreach {_.remove{taskIndex}}
-          stageAttemptToNumSpeculativeTasks(stageAttempt) -= 1
+          if (stageAttemptToNumSpeculativeTasks.contains(stageAttempt)) {
+            stageAttemptToNumSpeculativeTasks(stageAttempt) -= 1
+          }
```

Comment:
Since this is non-trivial, could you add some comments describing when this happens?

Reply:
Or, we can add some meaningful log message with `else` statement.

Reply:
comment added.

Reply:
Thanks!